File: MealBuilder.Domain/Recipes/RecipeNutrition.cs
namespace MealBuilder.Domain.Recipes;

public readonly record struct RecipeNutrition(
    decimal Calories,
    decimal Protein,
    decimal Fat,
    decimal Carbohydrates,
    decimal Sugars,
    decimal Fiber,
    decimal Salt)
{
    public static RecipeNutrition Zero { get; } = new(
        Calories: 0,
        Protein: 0,
        Fat: 0,
        Carbohydrates: 0,
        Sugars: 0,
        Fiber: 0,
        Salt: 0);

    public RecipeNutrition Add(RecipeNutrition nutrition)
    {
        return new RecipeNutrition(
            Calories + nutrition.Calories,
            Protein + nutrition.Protein,
            Fat + nutrition.Fat,
            Carbohydrates + nutrition.Carbohydrates,
            Sugars + nutrition.Sugars,
            Fiber + nutrition.Fiber,
            Salt + nutrition.Salt);
    }

    public RecipeNutrition MultiplyBy(decimal multiplier)
    {
        if (multiplier < 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(multiplier),
                multiplier,
                "Multiplier cannot be negative.");
        }

        return new RecipeNutrition(
            Calories * multiplier,
            Protein * multiplier,
            Fat * multiplier,
            Carbohydrates * multiplier,
            Sugars * multiplier,
            Fiber * multiplier,
            Salt * multiplier);
    }

    public RecipeNutrition DivideBy(int divisor)
    {
        return DivideBy((decimal)divisor);
    }

    public RecipeNutrition DivideBy(decimal divisor)
    {
        if (divisor <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(divisor),
                divisor,
                "Divisor must be greater than zero.");
        }

        return new RecipeNutrition(
            Calories / divisor,
            Protein / divisor,
            Fat / divisor,
            Carbohydrates / divisor,
            Sugars / divisor,
            Fiber / divisor,
            Salt / divisor);
    }
}

File: MealBuilder.Client/src/utils/recipeNutrition.ts
import type { Ingredient } from '../api/ingredientApi'
import type { RecipeNutrition } from '../api/recipeApi'

export type RecipeIngredientSelection = {
  ingredientId: number | null
  grams: number
}

export type RecipeNutritionSummary = {
  total: RecipeNutrition
  perServing: RecipeNutrition
}

function createZeroNutrition(): RecipeNutrition {
  return {
    calories: 0,
    protein: 0,
    fat: 0,
    carbohydrates: 0,
    sugars: 0,
    fiber: 0,
    salt: 0,
  }
}

function divideNutrition(
  nutrition: RecipeNutrition,
  divisor: number,
): RecipeNutrition {
  return {
    calories: nutrition.calories / divisor,
    protein: nutrition.protein / divisor,
    fat: nutrition.fat / divisor,
    carbohydrates: nutrition.carbohydrates / divisor,
    sugars: nutrition.sugars / divisor,
    fiber: nutrition.fiber / divisor,
    salt: nutrition.salt / divisor,
  }
}

export function calculateRecipeNutrition(
  selections: RecipeIngredientSelection[],
  ingredients: Ingredient[],
  servings: number,
): RecipeNutritionSummary {
  const ingredientsById = new Map(
    ingredients.map((ingredient) => [
      ingredient.id,
      ingredient,
    ]),
  )

  const total = selections.reduce<RecipeNutrition>(
    (nutrition, selection) => {
      if (
        selection.ingredientId === null ||
        !Number.isFinite(selection.grams) ||
        selection.grams <= 0
      ) {
        return nutrition
      }

      const ingredient = ingredientsById.get(
        selection.ingredientId,
      )

      if (!ingredient) {
        return nutrition
      }

      const multiplier = selection.grams / 100

      return {
        calories:
          nutrition.calories +
          ingredient.caloriesPer100g * multiplier,
        protein:
          nutrition.protein +
          ingredient.proteinPer100g * multiplier,
        fat:
          nutrition.fat +
          ingredient.fatPer100g * multiplier,
        carbohydrates:
          nutrition.carbohydrates +
          ingredient.carbohydratesPer100g * multiplier,
        sugars:
          nutrition.sugars +
          ingredient.sugarsPer100g * multiplier,
        fiber:
          nutrition.fiber +
          ingredient.fiberPer100g * multiplier,
        salt:
          nutrition.salt +
          ingredient.saltPer100g * multiplier,
      }
    },
    createZeroNutrition(),
  )

  const hasValidServings =
    Number.isFinite(servings) && servings > 0

  return {
    total,
    perServing: hasValidServings
      ? divideNutrition(total, servings)
      : createZeroNutrition(),
  }
}